### ISSUE TITLE
recode swallows fine unicode, but not plain HTML char entities for French characters?

### ISSUE THREAD

**Issue by Albretch** [rrthomas/recode#51]
_AZ="&#28608;&#20809;, &#36889;&#20841;&#20491;&#23383;&#26159;&#29978;&#40636;&#24847;&#24605;"
###
_AZ=$(echo "${_AZ}" | recode html..utf-8)
echo "// __ \$_AZ: |${_AZ}|"
// __ $_AZ: |激光, 這兩個字是甚麼意思|

_AZ="t%C3%AAte-%C3%A0-t%C3%AAte"
...
// __ $_AZ: |t%C3%AAte-%C3%A0-t%C3%AAte|
 it should be: "tête-à-tête"

How do you make recode give you UTF-8 regardless of the input string (which encoding should be easy to figure out based on the patterns of the input string)?

**Comment by rrthomas** [comment]
> How do you make recode give you UTF-8 regardless of the input string (which encoding should be easy to figure out based on the patterns of the input string)?

I don't see any HTML character entities. Your examples look like URL escaping, not HTML character entities.

> which encoding should be easy to figure out based on the patterns of the input string

Recode does not attempt to guess what encoding its input uses, it uses the encoding you tell it; you'd need another tool to guess encodings.

**Comment by Albretch** [comment]
OK, is there a way to make recode get as input: "t%C3%AAte-%C3%A0-t%C3%AAte"
 and give as output: "tête-à-tête" ?

**Comment by rrthomas** [comment]
No, I don't think recode supports URL encoding. That would be a good thing to add.

**Comment by KlfJoat** [comment]
I was looking to Recode to help me with this, as well.

FYI, the Wikipedia entry has changed to https://en.wikipedia.org/wiki/Percent-encoding since the technical term for this (keyword trapping in this paragraph) is not URLencoding, it is percent encoding. It was initially specified in RFC1738, but now is in [RFC3986](https://www.rfc-editor.org/rfc/rfc3986.html#section-2.1) (aka STD66).  

Maybe this is the kind of thing a Recode "surface" might be used for, at least for decoding. That was the first place I looked for it in the Recode docs. However, there are a few possibly different ways to decide what to encode, so maybe a normal charset would be best?

Most commonly, it is used for URIs to encode [reserved characters](https://www.rfc-editor.org/rfc/rfc3986.html#section-2.2) when they are not being used as delimiters. But it is also often used for any character that's not a reserved character or [unreserved character](https://www.rfc-editor.org/rfc/rfc3986.html#section-2.3). I can see those being two different encoding use cases.

Also, I believe it is used for IRIs ([RFC3987](https://www.rfc-editor.org/rfc/rfc3987)). When encoding, going [from UCS to UTF-8 to percent encoding](https://www.rfc-editor.org/rfc/rfc3987#section-3.1) (excepting the domain name part). This turns an IRI into an STD66 URI.

But there are, of course, plenty of example code libraries out there to help with deciding what and how to encode.

One test example I have...

```plaintext
file:///home/user/%D0%91%D0%AA%D0%9B%D0%93%D0%90%D0%A0%D0%A1%D0%9A%D0%98
```

should become...

```plaintext
file:///home/user/БЪЛГАРСКИ
```

(`/home/user/BULGARIAN` in Cyrillic)

**Comment by rrthomas** [comment]
Thanks for the detailed request. I would happily accept a patch for this.
